In [2]:
import pandas as pd
import numpy as np

from sklearn import datasets
data = datasets.load_diabetes()

from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [3]:
data.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

In [4]:
data['data'].shape

(442, 10)

In [5]:
data['target'].shape

(442,)

In [6]:
len(data['feature_names'])

10

In [10]:
df = pd.DataFrame(data['data'], index=data['target'], columns=data['feature_names'])
df

age       sex       bmi        bp        s1        s2        s3  \
151.0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
75.0  -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
141.0  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
206.0 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
135.0  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
...         ...       ...       ...       ...       ...       ...       ...   
178.0  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
104.0 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
132.0  0.041708  0.050680 -0.015906  0.017282 -0.037344 -0.013840 -0.024993   
220.0 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
57.0  -0.045472 -0.044642 -0.073030 -0.081414  0.083740  0.027809  0.173816   

             s4        s5        s6  
151.0 -0.002592  0.019908 -0.017646  
75.0  -0.039493 -0.068330 -0.092204  
141.0 -0.002592  0.002864 -0.025930  
206.0  0.034309  0.022692 -0.009362  
135.0 -0.002592 -0.031991 -0.046641  
...         ...       ...       ...  
178.0 -0.002592  0.031193  0.007207  
104.0  0.034309 -0.018118  0.044485  
132.0 -0.011080 -0.046879  0.015491  
220.0  0.026560  0.044528 -0.025930  
57.0  -0.039493 -0.004220  0.003064  

[442 rows x 10 columns]

단순 선형 회귀 분석
1. 회귀식의 추정
- 두 변수 X와 Y의 관계(bmi, target)에 적합한 회귀식을 구하기 위해서는 관측된 값으로부터 회귀계수 B0와 B1의 값을 추정하여야 합니다. 
- 이 때 일반적으로 많이 사용되는 방법을 최소제곱법이라고 합니다.

In [11]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [12]:
y = df.index.values
y

array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
        69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
        68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
        87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
       259.,  53., 190., 142.,  75., 142., 155., 225.,  59., 104., 182.,
       128.,  52.,  37., 170., 170.,  61., 144.,  52., 128.,  71., 163.,
       150.,  97., 160., 178.,  48., 270., 202., 111.,  85.,  42., 170.,
       200., 252., 113., 143.,  51.,  52., 210.,  65., 141.,  55., 134.,
        42., 111.,  98., 164.,  48.,  96.,  90., 162., 150., 279.,  92.,
        83., 128., 102., 302., 198.,  95.,  53., 134., 144., 232.,  81.,
       104.,  59., 246., 297., 258., 229., 275., 281., 179., 200., 200.,
       173., 180.,  84., 121., 161.,  99., 109., 115., 268., 274., 158.,
       107.,  83., 103., 272.,  85., 280., 336., 281., 118., 317., 235.,
        60., 174., 259., 178., 128.,  96., 126., 28

In [13]:
x = df.bmi.values
x

array([ 0.06169621, -0.05147406,  0.04445121, -0.01159501, -0.03638469,
       -0.04069594, -0.04716281, -0.00189471,  0.06169621,  0.03906215,
       -0.08380842,  0.01750591, -0.02884001, -0.00189471, -0.02560657,
       -0.01806189,  0.04229559,  0.01211685, -0.0105172 , -0.01806189,
       -0.05686312, -0.02237314, -0.00405033,  0.06061839,  0.03582872,
       -0.01267283, -0.07734155,  0.05954058, -0.02129532, -0.00620595,
        0.04445121, -0.06548562,  0.12528712, -0.05039625, -0.06332999,
       -0.03099563,  0.02289497,  0.01103904,  0.07139652,  0.01427248,
       -0.00836158, -0.06764124, -0.0105172 , -0.02345095,  0.06816308,
       -0.03530688, -0.01159501, -0.0730303 , -0.04177375,  0.01427248,
       -0.00728377,  0.0164281 , -0.00943939, -0.01590626,  0.0250506 ,
       -0.04931844,  0.04121778, -0.06332999, -0.06440781, -0.02560657,
       -0.00405033,  0.00457217, -0.00728377, -0.0374625 , -0.02560657,
       -0.02452876, -0.01806189, -0.01482845, -0.02991782, -0.04

In [14]:
# 2d array로 바꾸기
x = x.reshape(-1,1)
y = y.reshape(-1,1)

In [15]:
lr.fit(x, y)

LinearRegression()

In [19]:
print(lr.coef_[0])
print(lr.intercept_)
# x를 먼저 넣고 y를 차례대로 변수로 입력해주고 실행하면 끝
# 결과를 보면 회귀 계수를 구할 수 있고  
# 여기서는 B1 = 949.435, B0=152.133
# Target = 949*bmi + 152
# 타켓의 수치가 1이 증가할 때마다 bmi가 949만큼 증가함

[949.43526038]
[152.13348416]


2. 회귀모형의 검정 및 적합도 파악

 이 회귀식이 통계적으로 유의한지, 변수가 유의하게 영향을 미치는 지, 그리고 얼만큼의 설명력을 가지는지 등의 여부를 확인해야 합니다.

A. F-statistic

도출된 회귀식이 회귀분석 모델 전체에 대해 통계적으로 의미가 있는지 파악

B. P-Value

각 변수가 종속변수에 미치는 영향이 유의한지 파악

C. 수정된 R제곱

회귀직선에 의하여 설명되는 변동이 총변동 중에서 차지하고 있는 상대적인 비율이 얼마인지 나타냄

즉, 회귀직선이 종속변수의 몇%를 설명할 수 있는지 확인

In [21]:
import statsmodels.api as sm
results = sm.OLS(y, sm.add_constant(x)).fit()

In [22]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.344
Model:                            OLS   Adj. R-squared:                  0.342
Method:                 Least Squares   F-statistic:                     230.7
Date:                Mon, 24 Jul 2023   Prob (F-statistic):           3.47e-42
Time:                        12:30:06   Log-Likelihood:                -2454.0
No. Observations:                 442   AIC:                             4912.
Df Residuals:                     440   BIC:                             4920.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.974     51.162      0.000     146.289     157.978
x1           949.4353     62.515     15.187      0.000     826.570    1072.301
==============================================================================
Omnibus:                       11.674   Durbin-Watson:                   1.848
Prob(Omnibus):                  0.003   Jarque-Bera (JB):                7.310
Skew:                           0.156   Prob(JB):                       0.0259
Kurtosis:                       2.453   Cond. No.                         21.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

A. F-statistic의 p-value 값은 Prob(F-statistic)으로 표현되는데, 이는 3.47e-42로 0.05보다 작기에 이 회귀식은 회귀분석 모델 전체에 대해 통계적으로 의미가 있다고 볼 수 있습니다.

B. 중간쯤에 보면 coef와 변수 x1의 p-value 값이 나와있습니다. 여기서 x1은 bmi이고 이 변수의 p-value가 0.000으로 표기 되어 있기에 0.05보다 작으므로 target을 설명하는데 유의하다고 판단할 수 있습니다.

C. 제일 위 부분에 R-squared와 Adj. R-squared가 표기되어 있는데, 값이 0.34정도로 이는 34%만큼의 설명력을 가진다고 판단할 수 있습니다. 참고로, 0에 가까울 수록 예측값을 믿을 수 없고 1에 가까울 수록 믿을 수 있다고 보면 됩니다.

 

중선형 회귀분석

###### 단순선형회귀분석과 비교했을 떄 종속변수를 설명하는 독립변수가 두 개 이상으로 증가함

In [25]:
x = df.loc[:, ['bmi', 'age', 'sex']].values
x

array([[ 0.06169621,  0.03807591,  0.05068012],
       [-0.05147406, -0.00188202, -0.04464164],
       [ 0.04445121,  0.08529891,  0.05068012],
       ...,
       [-0.01590626,  0.04170844,  0.05068012],
       [ 0.03906215, -0.04547248, -0.04464164],
       [-0.0730303 , -0.04547248, -0.04464164]])

In [27]:
import statsmodels.api as sm
results = sm.OLS(y, sm.add_constant(x)).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.351
Model:                            OLS   Adj. R-squared:                  0.346
Method:                 Least Squares   F-statistic:                     78.94
Date:                Mon, 24 Jul 2023   Prob (F-statistic):           7.77e-41
Time:                        12:37:42   Log-Likelihood:                -2451.6
No. Observations:                 442   AIC:                             4911.
Df Residuals:                     438   BIC:                             4928.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.964     51.321      0.000     146.307     157.960
x1           926.9120     63.525     14.591      0.000     802.061    1051.763
x2           138.9039     64.254      2.162      0.031      12.618     265.189
x3           -36.1353     63.391     -0.570      0.569    -160.724      88.453
==============================================================================
Omnibus:                       14.687   Durbin-Watson:                   1.851
Prob(Omnibus):                  0.001   Jarque-Bera (JB):                8.290
Skew:                           0.150   Prob(JB):                       0.0158
Kurtosis:                       2.400   Cond. No.                         23.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- 회귀식은 Target = 926*bmi + 138*age -36*sex + 152로 도출된다.
- F-statistic의 p-value 값은 Prob(F-statistic)으로 표현되는데, 이는 7.77e-41로 0.05보다 작으므로 이 회귀식은 회구분석 모델 전체에 대해 통계적으로 의미가 있다.
- 중간쯤에 보면 coef와 여러 변수 들의 p-value 값이 나와 있다. 여기서 x1, x2는 변수의 p-value가 0.05미만으로 유의미하다고 볼 수 있다. x3의 p-value는 0.569로 0.05보다 크게이 유의미하다고 판단할 수 없다. 
- 즉, sex는 target을 설명하는데 유의하다고 판단할 수 없다.
- 제일 위 부분에 R-squared와 Adj.R-squared가 표기되어 있는데, 값이 0.35정도로 이는 35%만큼의 설명력을 가지고 있다고 판단된다.